sample-prep/Biomek_pooling_VI.py

```diff
@@ -29,107 +43,287 @@ def main(process_id, file_id):
+        source_location.append(input.location)
+        sample_name.append(input.name)
+        molarity.append(input.udf.get('Molarity TS (nM)'))
+        project_account.append(input.udf.get('Project Account'))
+        genome_size.append(input.udf.get('Genome Size (Mbp)'))
+        species.append(input.udf.get('Species'))
+        input_container_type.append(input.container.type.name)
+        input_container_name.append(input.container.name)
+        sample_volumes.append(round(molMbp*input.udf.get('Genome Size (Mbp)')/input.udf.get('Molarity TS (nM)'),2))
 
-    #exit script if entered step with more than one pool
+    # loop for 'Enforce Min Volume Required' button on the step. 
+    # If true, find the minimum value, find its index.
+    # Recalculate molMbp to achieve min_vol_required. New value put to the process udf for info to the user. volume_threshold_plotcolor set to yellow
+    # finally a for-loop to update sample_volumes using the new fmol/Mbp
 
-    unique_pools_dict = dict(zip(list(Pool_names),[list(Pool_names).count(i) for i in list(Pool_names)]))
-    Pool_names_unique = (list(unique_pools_dict.keys()))
+    if enforce_min_volume == True:
+        min_volume_in_list = min(sample_volumes)
+        min_volume_in_list_index = sample_volumes.index(min_volume_in_list)
+        if min_volume_in_list < min_vol_required: 
+            molMbp = (min_vol_required*molarity[min_volume_in_list_index])/genome_size[min_volume_in_list_index]
+            process.udf['fmol/Mbp'] = molMbp
+            process.put()
 
-    if len(Pool_names_unique) > 1:
-        print("This step only supports one pool at the time. Return to ice bucket and repeat with a single pool")
-        sys.exit(1)
+            index = 0
+            for input in sorted(inputs, key=sort_key):
+                sample_volumes[index] = (round(molMbp*input.udf.get('Genome Size (Mbp)')/input.udf.get('Molarity TS (nM)'),2))
+                index += 1
+
+    # volume_threshold_plotcolor to allow conditional coloring on plot
+    # adjusting volume to max_vol_available if enforce_max_vol
+    # over or under threshold = red, within threshold = palegreen, adjusted volume to max = yellow 
+
+    volume_threshold_plotcolor = []
+
+    for volumes in sample_volumes:
+        if volumes > max_vol_available or volumes < min_vol_required:
+            volume_threshold_plotcolor.append("red")
+        else:
+            volume_threshold_plotcolor.append("palegreen") 
+
+    if enforce_max_vol == True:
+        for i in range(len(sample_volumes)):
+            if sample_volumes[i] > max_vol_available:
+                sample_volumes[i] = max_vol_available
+                volume_threshold_plotcolor[i] = "yellow"
 
-    #write csv-file
+    if enforce_min_volume == True:
+        if min_volume_in_list < min_vol_required:
+            volume_threshold_plotcolor[min_volume_in_list_index] = "yellow"
+
+    # DEFINE BIOMEK DECK PLATE POSITIONS (LIBLABWARE)
```

Comment:
Is Liblabware a term that a lot of labfolks know of?

Reply:
Liblabware is a term from the biomek robot. So not a general term, just a biomek term. Here, Rosa has told the robot that Liblabware "Libraries1" and "Libraries2" are two different plates with defined positions on the deck. When running the robot you are told where to place them. And the csv-file will tell the robot which volumes it need to collect from those plates.